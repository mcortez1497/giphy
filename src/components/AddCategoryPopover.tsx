import React, { ChangeEvent, KeyboardEvent } from "react";

import {
  Button,
  IconButton,
  InputAdornment,
  Paper,
  TextField,
  WithStyles
} from "@material-ui/core";
import { createStyles, Theme, withStyles } from "@material-ui/core/styles";
import { Add } from "@material-ui/icons";

const styles = (theme: Theme) =>
  createStyles({
    root: {
      padding: theme.spacing(0, 1, 1, 1)
    }
  });

interface Props extends WithStyles<typeof styles> {
  readonly onClose: () => void;
  readonly onSubmit: (categoryName: string) => void;
}

interface State {
  categoryName: string;
}

class AddCategoryPopoverWithStyles extends React.Component<Props, State> {
  state = {
    categoryName: ""
  };

  render() {
    const {
      handleCategoryNameChange,
      handleSubmit,
      handleSubmitOnKeypress,
      props: { classes },
      state: { categoryName }
    } = this;

    const isValid = categoryName !== "";

    return (
      <div className={classes.root}>
        <TextField
          variant="outlined"
          margin="normal"
          required
          fullWidth
          id="category"
          label="Category Name"
          name="category"
          value={categoryName}
          onChange={handleCategoryNameChange}
          onKeyPress={handleSubmitOnKeypress}
          autoFocus
          InputProps={{
            endAdornment: (
              <InputAdornment position="end">
                <IconButton
                  disabled={!isValid}
                  edge="end"
                  aria-label="add category"
                  onClick={handleSubmit}
                >
                  <Add color={isValid ? "primary" : "disabled"} />
                </IconButton>
              </InputAdornment>
            )
          }}
        />
      </div>
    );
  }

  private handleSubmit = () => {
    if (this.state.categoryName === "") {
      return;
    }
    this.props.onSubmit(this.state.categoryName);
    this.props.onClose();
  };

  private handleSubmitOnKeypress = (event: KeyboardEvent<HTMLDivElement>) => {
    if (event.key === "Enter") {
      this.handleSubmit();
    }
  };

  private handleCategoryNameChange = (event: ChangeEvent<HTMLInputElement>) => {
    this.setState({ categoryName: event.target.value });
  };
}

const AddCategoryPopover = withStyles(styles)(AddCategoryPopoverWithStyles);

export { AddCategoryPopover };
