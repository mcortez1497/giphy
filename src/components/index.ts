export * from './Content';
export * from './ContentContainer';
export * from "./GifCard";
export * from "./GifCardContainer";
export * from "./Header";
export * from './HeaderContainer';
export * from "./LoginPopover";
export * from "./RegistrationPopover";
export * from "./SearchBar";
export * from "./UserDrawer";
export * from "./UserDrawerContainer";