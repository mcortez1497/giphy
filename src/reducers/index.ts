import { applyMiddleware, combineReducers, createStore } from "redux";
import { composeWithDevTools } from "redux-devtools-extension";
import reduxThunk from "redux-thunk";

import { gifReducer, GifState } from "./Gifs";
import { userReducer, UserState } from "./User";
import { uiReducer, UIState } from "./UI";

export * from "./Gifs";
export * from "./User";
export * from "./UI";

const middlewares = [reduxThunk];

export const store = createStore(
  combineReducers({
    gifs: gifReducer,
    user: userReducer,
    ui: uiReducer
  }),
  composeWithDevTools(applyMiddleware(...middlewares))
);

export interface AppState {
  gifs: GifState;
  user: UserState;
  ui: UIState;
}
