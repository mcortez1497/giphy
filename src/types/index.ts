export * from "./category";
export * from "./gif";
export * from "./util";