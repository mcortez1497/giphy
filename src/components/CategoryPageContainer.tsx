import * as React from "react";
import { connect } from "react-redux";
import { RouteComponentProps } from "react-router";
import { bindActionCreators, Dispatch } from "redux";

import { Content, Layout } from "components";
import { AppState, getGifs } from "reducers";
import { Gif } from "types";

interface StateProps {
  readonly gifs: Gif[];
}

interface DispatchProps {
  readonly getGifs: () => void;
}

interface RouteProps {
  readonly categoryId: string;
}

interface ComponentProps
  extends StateProps,
    DispatchProps,
    RouteComponentProps<RouteProps> {}

class Container extends React.Component<ComponentProps> {
  public componentDidMount() {
    this.props.getGifs();
  }

  public render() {
    return (
      <Layout>
        <Content gifs={this.props.gifs} />
      </Layout>
    );
  }
}

const mapStateToProps = (
  state: AppState,
  props: RouteComponentProps<RouteProps>
) => ({
  gifs: state.user.gifs.reduce(
    (gifsWithCategory: Gif[], gif: Gif) =>
      gif.categories &&
      gif.categories.find(
        category => category._id === props.match.params.categoryId
      )
        ? [...gifsWithCategory, gif]
        : gifsWithCategory,
    []
  )
});

const mapDispatchToProps = (dispatch: Dispatch) =>
  bindActionCreators(
    {
      getGifs
    },
    dispatch
  );

export const CategoryPageContainer = connect<
  StateProps,
  DispatchProps,
  RouteComponentProps<RouteProps>,
  AppState
>(
  mapStateToProps,
  mapDispatchToProps
)(Container);
