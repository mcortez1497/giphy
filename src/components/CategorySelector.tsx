import React, { ChangeEvent } from "react";

import {
  Chip,
  FormControl,
  Input,
  InputLabel,
  MenuItem,
  Select,
  WithStyles
} from "@material-ui/core";
import { createStyles, Theme, withStyles } from "@material-ui/core/styles";

import { Category, Gif } from "types";

const styles = (theme: Theme) =>
  createStyles({
    chips: {
      display: "flex",
      flexWrap: "wrap"
    },
    chip: {
      margin: 2
    },
    grow: {
      flexGrow: 1
    },
    label: {
      top: "-7px"
    },
    select: {
      flexGrow: 1,
      "&::before": {
        borderWidth: 0
      }
    }
  });

const MenuProps = {
  PaperProps: {
    style: {
      maxHeight: 48 * 4.5 + 8,
      width: 250
    }
  }
};

interface Props extends WithStyles<typeof styles> {
  readonly categories: Category[];
  readonly gif: Gif;

  readonly onCategorySelect: (gifId: string, categories: string[]) => void;
}

const CategorySelectorWithStyles: React.FC<Props> = ({
  categories,
  classes,
  gif,
  onCategorySelect
}) => (
  <FormControl className={classes.grow}>
    {!gif.categories ||
      (gif.categories.length <= 0 && (
        <InputLabel htmlFor="select-multiple-chip" className={classes.label}>
          Add Categories
        </InputLabel>
      ))}
    <Select
      multiple
      value={getSelectedCategories(gif)}
      onChange={handleCategorySelect(gif, onCategorySelect)}
      className={classes.select}
      input={<Input id="select-multiple-chip" />}
      renderValue={selected => (
        <div className={classes.chips}>
          {(selected as string[]).map(categoryId => (
            <Chip
              key={categoryId}
              label={getCategoryNameById(categoryId, gif.categories)}
              className={classes.chip}
            />
          ))}
        </div>
      )}
      MenuProps={MenuProps}
      IconComponent={() => null}
    >
      >
      <MenuItem value="" disabled>
        Add Categories
      </MenuItem>
      {categories.map((category, index) => (
        <MenuItem
          key={index}
          value={category._id}
        >
          {category.name}
        </MenuItem>
      ))}
    </Select>
  </FormControl>
);

const getSelectedCategories = (gif: Gif) =>
  gif.categories ? gif.categories.map(cat => cat._id) : [];

const getCategoryNameById = (
  categoryId: string,
  categories: Category[] = []
) => {
  const category = categories.find(category => category._id === categoryId);
  return category ? category.name : "";
};

const handleCategorySelect = (gif: Gif, onCategorySelect: Function) => (
  event: ChangeEvent<{ name?: string | undefined; value: any }>
) => {
  if (!gif._id) {
    return;
  }
  onCategorySelect(gif._id, event.target.value);
};

const CategorySelector = withStyles(styles)(CategorySelectorWithStyles);

export { CategorySelector };
