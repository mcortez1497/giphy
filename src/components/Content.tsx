import React from "react";
import { GifViewer } from "../components";
import { Gif } from "../types";

interface Props {}

interface State {
  gifs: Gif[];
}

class Content extends React.Component<Props, State> {
  state = {
    gifs: []
  };

  async componentDidMount() {
    fetch(
      "http://api.giphy.com/v1/gifs/trending?api_key=osVl8xeSb13OvfguXaI0pUrbRzph61v9"
    )
      .then(response => response.json())
      .then(json => {
        const gifs = json.data.map((gif: Gif) => gif);
        this.setState({
          gifs
        });
      })
      .catch(error => console.log(error));
  }

  render() {
    return (
      <div>
        <div>
          <input type="text" placeholder="Search for GIFs" />
        </div>
        <GifViewer gifs={this.state.gifs} />
      </div>
    );
  }
}

export { Content };
