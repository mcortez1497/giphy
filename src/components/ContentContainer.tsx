import * as React from 'react';
import { connect } from 'react-redux';
import { bindActionCreators, Dispatch } from 'redux';

import { AppState, getGifs } from 'reducers';
import { Gif } from 'types';

import { Content } from 'components';

interface StateProps {
  readonly gifs: Gif[];
}

interface DispatchProps {
  readonly getGifs: () => void;
}

interface ComponentProps extends StateProps, DispatchProps {}

class Container extends React.Component<ComponentProps> {
  public componentDidMount() {
    this.props.getGifs();
  }

  public render() {
    return <Content gifs={this.props.gifs} />;
  }
}

const mapStateToProps = (state: AppState) => ({
  gifs: state.gifs.items
})

const mapDispatchToProps = (dispatch: Dispatch) =>
  bindActionCreators(
    {
      getGifs
    },
    dispatch
  );

export const ContentContainer = connect<StateProps, DispatchProps, {}, AppState>(
  mapStateToProps,
  mapDispatchToProps
)(Container);
