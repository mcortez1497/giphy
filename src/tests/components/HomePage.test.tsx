import { TestHelper } from "tests";

import { HomePage } from "components";
import { defaultApiRequest } from "reducers";
import { Gif } from "types";

const props = {
  apiRequest: defaultApiRequest,
  gifs: [] as Gif[],
  query: "",
  getGifs: jest.fn(),
  getMoreGifs: jest.fn()
};

describe("HomePage", () => {
  it("renders correctly", () => {
    const { element } = TestHelper.getInstanceObjects(HomePage, props);
    expect(element).toMatchSnapshot();
  });

  it("renders without InfiniteScroll if in error state", () => {
    const { element } = TestHelper.getInstanceObjects(HomePage, {
      ...props,
      apiRequest: {
        ...props.apiRequest,
        isError: true
      }
    });
    expect(element).toMatchSnapshot();
  });

  it("fetches data on componentDidMount", () => {
    const query = "test";

    TestHelper.getInstanceObjects(
      HomePage,
      { ...props, query },
      {
        disableLifecycleMethods: false
      }
    );

    expect(props.getGifs).toHaveBeenCalledWith(query);
  });
});
