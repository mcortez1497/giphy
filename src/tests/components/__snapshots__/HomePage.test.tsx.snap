// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HomePage renders correctly 1`] = `
<Layout>
  <WithWidth(WithStyles(ContentWithStyles))
    allowTrim={true}
    apiRequest={
      Object {
        "isError": false,
        "isLoading": false,
        "isSuccess": false,
      }
    }
    gifs={Array []}
    isAuthenticated={true}
    showEmptyMessage={false}
  />
  <ForwardRef(WithStyles)
    loadMore={[MockFunction]}
  />
</Layout>
`;

exports[`HomePage renders without InfiniteScroll if in error state 1`] = `
<Layout>
  <WithWidth(WithStyles(ContentWithStyles))
    allowTrim={true}
    apiRequest={
      Object {
        "isError": true,
        "isLoading": false,
        "isSuccess": false,
      }
    }
    gifs={Array []}
    isAuthenticated={true}
    showEmptyMessage={false}
  />
</Layout>
`;
